src/fontcache.cpp

```diff
@@ -1056,9 +1056,301 @@ static void LoadWin32Font(FontSize fs)
 	new Win32FontCache(fs, logfont, settings->size);
 }
 
+#elif defined(WITH_COCOA)
+
+#include "os/macosx/macos.h"
+#include <CoreFoundation/CoreFoundation.h>
+#include <cmath>
+
+class CoreTextFontCache : public TrueTypeFontCache {
+	CFAutoRelease<CTFontDescriptorRef> font_desc; ///< Font descriptor exlcuding font size.
+	CFAutoRelease<CTFontRef> font;                ///< CoreText font handle.
+
+	std::string font_name;                        ///< Cached font name.
+
+	void SetFontSize(int pixels);
+	const Sprite *InternalGetGlyph(GlyphID key, bool aa) override;
+	const void *InternalGetFontTable(uint32 tag, size_t &length) override;
+public:
+	CoreTextFontCache(FontSize fs, CFAutoRelease<CTFontDescriptorRef> &&font, int pixels);
+	~CoreTextFontCache() {}
+
+	void ClearFontCache() override;
+	GlyphID MapCharToGlyph(WChar key) override;
+	const char *GetFontName() override { return font_name.c_str(); }
+	bool IsBuiltInFont() override { return false; }
+	const void *GetOSHandle() override { return font.get(); }
+};
+
+CoreTextFontCache::CoreTextFontCache(FontSize fs, CFAutoRelease<CTFontDescriptorRef> &&font, int pixels) : TrueTypeFontCache(fs, pixels), font_desc(std::move(font))
+{
+	this->SetFontSize(pixels);
+}
+
+/**
+ * Reset cached glyphs.
+ */
+void CoreTextFontCache::ClearFontCache()
+{
+	/* GUI scaling might have changed, determine font size anew if it was automatically selected. */
+	if (this->font) this->SetFontSize(this->req_size);
+
+	this->TrueTypeFontCache::ClearFontCache();
+}
+
+void CoreTextFontCache::SetFontSize(int pixels)
+{
+	if (pixels == 0) {
+		/* Try to determine a good height based on the height recommended by the font. */
+		int scaled_height = ScaleFontTrad(_default_font_height[this->fs]);
+		pixels = scaled_height;
+
+		CFAutoRelease<CTFontRef> f(CTFontCreateWithFontDescriptor(this->font_desc.get(), 0.0f, nullptr));
+		if (f) {
+			float min_size = 0.0f;
+
+			/* The 'head' TrueType table contains information about the
+			 * 'smallest readable size in pixels'. Try to read it out. If
+			 * that doesn't work, we use the default OS font size instead. */
+			CFAutoRelease<CFDataRef> data(CTFontCopyTable(f.get(), kCTFontTableHead, kCTFontTableOptionNoOptions));
+			if (data) {
+				uint16_t lowestRecPPEM; // At offset 46 of the 'head' TrueType table.
+				CFDataGetBytes(data.get(), CFRangeMake(46, sizeof(lowestRecPPEM)), (UInt8 *)&lowestRecPPEM);
+				min_size = CFSwapInt16BigToHost(lowestRecPPEM); // TrueType data is always big-endian.
+			} else {
+				CFAutoRelease<CFNumberRef> size((CFNumberRef)CTFontCopyAttribute(f.get(), kCTFontSizeAttribute));
+				CFNumberGetValue(size.get(), kCFNumberFloatType, &min_size);
+			}
+
+			/* Font height is minimum height plus the difference between the default
+			 * height for this font size and the small size. */
+			int diff = scaled_height - ScaleFontTrad(_default_font_height[FS_SMALL]);
+			pixels = Clamp(std::min(min_size, 20.0f) + diff, scaled_height, MAX_FONT_SIZE);
```

Comment:
`20.0f` <- magic value; please make a const out of it or comment where it comes from